### PR TITLE
Webpack v5; Use PostCSS v8 with Tailwind Setup

### PR DESCRIPTION
**Pull request by thedavidprice**
_continues work from #2402 after package dependency re-structure_

## Project
- [x] `yarn rw build` including prerender
- [x] `yarn rw dev`
- [x] build and dev with `--esbuild`
- [x] `yarn rw serve`
- [x] Storybook: runs and builds successfully (update: upon re-link'ing, both now failing)
    - NOTE: Storybook v6.3 upgrade work was completed here https://github.com/redwoodjs/redwood/pull/2892
- [x] yarn rw test

## CI
- [x] `yarn build`
- [x] `yarn lint`
- [x] `yarn test`
- [x] E2E
    - [x] tutorial (both original and envelop/helix)
    - [x] cli: _build is failing_
    - [x] storybook
    - [x] logger

# Release Notes
**Tailwind:** Provide instructions to use PostCSS v8 with Tailwind
- [ ] can we enable purge by default?

### CODE REVIEW

File: packages/testing/config/storybook/main.js
Comment (thedavidprice) [inline packages/testing/config/storybook/main.js:57]: @peterp I could use some help debugging this one. I get an "undefined" error when trying to run or build storybook if this is enabled.

Comment (thedavidprice) [reply packages/testing/config/storybook/main.js:57]: Scratch that. After re-linking my project, I can no longer get Storybook to build or run.

**Comment by thedavidprice** [comment]
@peterp @dac09 
This one's in really good shape and ready for review. I need help resolving the CI issue below before merging. "Error Overlay" is mostly FYI — open to options/suggestions if any.

### Failing E2E on CI
Locally, E2E passes with flying colors.

However, after multiple runs, I cannot pass rw build or storybook:
- [CLI build command failures](https://github.com/redwoodjs/redwood/pull/2884/checks?check_run_id=3069067589#step:9:93)
- [Storybook builds correctly but stories cannot be accessed](https://github.com/redwoodjs/redwood/pull/2884/checks?check_run_id=3069067589#step:9:213)

I have no idea how to diagnose/debug.

If we proceed, we'll continue to fail all CI E2E I fear 😢

### Disabled Error Overlay

The new DevServer v4 has the built in React error overlay enabled by default. It's not spectacular, but it works (screenshot at bottom).

I've disabled the React-refresh overlay in webpack.common.js, which I don't believe we needed anyway as it was redundant:
```
  new ReactRefreshWebpackPlugin({ overlay: false }), // bug with Webpack v5
```

I also had to disable error-overlay-webpack-plugin in webpack.development.js, which is a bummer because it is an improvement compared to the default display. Currently I've just uncommented the lines of code. We should discuss if we want to leave it as is or wait for v5 support or just copy+paste the plugin into our project and update the API changes (I know what they are now). The project is sitting on a multi-month open PR with no maintainer response. Not great.

#### Default React Error Overlay Screenshot
Screen Shot 2021-05-19 at 6 22 01 PM

**Comment by thedavidprice** [comment]
@peterp @dac09 
This is passing CI and ready to merge! Good news / Bad news — I think we've had some bugs all along that this upgrade helped to weed out. Also, I possibly had a mixed experience with `rwfw project:sync` --> some things I got working locally were _actual_ failures when run on GitHub CI. Might have had to do with yarn install resolutions — maybe we need to always run `yarn -f` in the context of rwfw usage?

@jtoar 
Please verify the Tailwind setup changes here. And also make any updates if needed to the default config. E.g. can we enable purge by default?

**Comment by thedavidprice** [comment]
@dac09 Dom and I previously discussed included your RFC https://github.com/redwoodjs/redwood/issues/1727 in this PR. I want to keep it on the radar, but unless there's an obvious quick implementation I suggest we hold until another time.

**Comment by peterp** [comment]
|  I think we've had some bugs all along that this upgrade helped to weed out. Also, I possibly had a mixed experience with rwfw project:sync --> some things I got working locally were actual failures when run on GitHub CI.

@thedavidprice Could you try create a reproduction?🤞 Would love to iron out all these issues!

**Comment by thedavidprice** [comment]
@peterp Will do. Mostly I'll be watching this ongoing and attempt to reproduce. The thing about inconsistent behavior is it's hard to roll back to diagnose. My best guess is that the packages installed in my project's node_modules weren't completely updated with `yarn install` during rwfw sync. I'll keep a look out!

To be clear, this PR is ready to go regardless.

**Comment by dac09** [comment]
LGTM @thedavidprice - but we should note this is likely a breaking change for anyone:
a) Using tailwind setup from before
b) Have custom webpack config

**Review by dac09 — APPROVED** [review]

File: packages/core/config/webpack.common.js
Comment (peterp) [inline packages/core/config/webpack.common.js:327]: Super weird syntax!

File: packages/core/package.json
Comment (peterp) [inline packages/core/package.json:49]: It's weird to me that they bumped some version beyond 5, but not all of them 🤷‍♂️

File: packages/testing/package.json
Comment (peterp) [inline packages/testing/package.json:36]:
```suggestion
```

**Review by peterp — APPROVED** [review]

File: packages/core/config/webpack.common.js
Comment (dac09) [reply packages/core/config/webpack.common.js:327]: haha yeah :D

File: packages/core/package.json
Comment (thedavidprice) [reply packages/core/package.json:49]: It's infuriating, actually. For this case re: `webpack-dev-server`, v3 is _technically_ webpack 5 compatible, however there's a known bug so you have to use the v4-beta. This means there are no docs for the breaking changes in dev-server v4, you have to dig through the changelogs yourself. Pain. Ful.

I was talking to Tom about all this today (just giving him an update). I don't really have a problem with Webpack. And I truly think v5 is an improvement. And even more so now I see the value Redwood adds because it internally handles these things. BUT I think in the future we have an opportunity to better manage (and even constrain) our own use of Webpack. E.g. the current changes to `yarn rw dev` is a great de-coupling choice. Also, we currently have too many potential layers of extended config, each including conditionals, that make it challenging to understand what Webpack is doing in a given context (and therefore what plugins/config are being applied). tl;dr: I think we'll be happy about Webpack if we:
- use it less
- simplify (and bound) our use

Thinking out loud here as it's fresh. Not suggesting we do any of this prior to v1.

**Comment by jtoar** [comment]
> LGTM @thedavidprice - but we should note this is likely a breaking change for anyone:
a) Using tailwind setup from before

@dac09 @thedavidprice I don't think this is actually breaking for postcss-v7-compatible tailwind setups (i.e. the setup command as it is now). I just tried it and it worked like before.

We'll provide an upgrade guide for getting current rw tailwind setups to the latest version of tailwind, postcss, autoprefixer, but it seems like they can also keep using their setup as it is if they want. Less overhead to upgrade their version of redwood for sure!

**Pull request merged**
